File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack = []
        arr = [None for _ in range(len(temperatures))]
        for i in range(len(temperatures)-1, -1, -1):
            while len(stack)>0 and stack[-1][0] <= temperatures[i]:
                stack.pop()
            if len(stack) == 0:
                arr[i] = 0
            else:
                arr[i] = stack[-1][1] - i
            stack.append([temperatures[i], i])
        return arr
        

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<pair<int, int>> sk;
        int n = temperatures.size();
        vector<int> sol(n);
        for(int i=n-1; i>=0; i--){
            while(!sk.empty() && sk.top().first <= temperatures[i])
                sk.pop();
            
            if(sk.empty()){
                sol[i] = 0;
            }
            else{
                sol[i] = sk.top().second - i;
            }
            
            sk.push({temperatures[i], i});
        }
        
        return sol;
    }
};